Niak 55 Spring security

Sewerinius opened the pull request:
Added Spring security to the backend.
Secured all the endpoints with 4 roles: 
"ROLE_ANON", "ROLE_STANDARD", "ROLE_SUPERVISOR", "ROLE_ADMINISTRATOR"
All endpoints are forbidden by default
There is an "/login" endpoint which you can use to log in
- The first user needs to be added to the database manually (you can use something like "{noop}pass" for password)

Sewerinius committed:
[NIAK-55] Added spring-security dependency

Sewerinius committed:
[NIAK-55] Better docker gradle caching

Sewerinius committed:
[NIAK-55] Added Spring Security dependency

Sewerinius committed:
[NIAK-55] Added Spring Security configuration

Sewerinius committed:
[NIAK-55] Updated storing and loading users

Sewerinius committed:
[NIAK-55] Secured all enpoints

Sewerinius committed:
[NIAK-55] Disabled the DesignForExtension check

Sewerinius committed:
[NIAK-55] Set Wildcard import check to warning level

Sewerinius committed:
[NIAK-55] Added Spring Security Test and fixed one test

Sewerinius committed:
[NIAK-55] Merge remote-tracking branch 'remotes/origin/develop' into NIAK-55

Sewerinius commented:
Now that I did something on the backend I have something to say about all those Request and Response classes.
As somebody correctly put them in a rest package: they are part of the rest logic so they shouldn't inside of services (which are supposed to do business logic). 
I think we should only use them to encapsulate objects inside of the Controller classes.

Sewerinius committed:
[NIAK-55] A little refactoring

enthyp commented:
Some conflicts appeared, I'll try to check this out tomorrow

Sewerinius committed:
[NIAK-55] Merge remote-tracking branch 'remotes/origin/develop' into NIAK-55

# Conflicts:
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentPeriodController.java
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/SystemUserController.java
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/service/CourseOfStudyService.java
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/service/UserService.java

Sewerinius committed:
[NIAK-55] Secured newest Controllers

Sewerinius committed:
Fixed permission problems

enthyp commented on rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/SystemUserController.java:
```diff
@@ -53,27 +61,41 @@ public GetSystemUsersResponse getSystemUsersByFirstNameAndLastName(@RequestParam
+    public Object getMe() {
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        System.out.println(auth);
+
```
I think Sout should be gone

enthyp commented on rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentController.java:
```diff
@@ -20,24 +21,27 @@
     private final RecruitmentService recruitmentService;
 
+    @Secured({"ROLE_STANDARD", "ROLE_SUPERVISOR", "ROLE_ADMINISTRATOR"})
     @PostMapping
```
Not sure if we want standard users to add recruitment?

enthyp commented on rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentController.java:
```diff
@@ -20,24 +21,27 @@
     @ResponseBody
     @ResponseStatus(value = HttpStatus.CREATED)
     public AddRecruitmentResponse addSystemUser(@RequestBody final AddRecruitmentRequest request) {
         return recruitmentService.saveRecruitment(request);
```
I know it's not the subject of this PR but we could use the opportunity and fix the names of methods in this class, it's related to recruitments not system users

enthyp commented on rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/service/UserService.java:
```diff
@@ -100,21 +102,32 @@ public void validateAddSystemUserRequest(final AddSystemUserRequest request) {
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        log.debug(username + " tries to log in");
+        try {
```
I think debug messages should be gone (here and in catch block)

enthyp commented on rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/model/security/UserPrincipal.java:
```diff
@@ -0,0 +1,50 @@
+
+@AllArgsConstructor
+public final class UserPrincipal implements UserDetails {
+    private final SystemUser user;
```
Do we need this class or will we need it later? It seems we can just return an instance of org.springframework.security.core.userdetails.User class in UserService.loadUserByUsername method

enthyp reviewed (commented):
I've added some minor remarks to the code. 

And also:
1. I encountered an issue: I added a user 'root' with password 'pass' to the DB and submitted, got a server error: 
```
rezerwacje-app_1    | java.lang.IllegalArgumentException: There is no PasswordEncoder mapped for the id "null"
```
What am I doing wrong?

2. Merging this PR alone would make working with front-end a pain until NIAK-61 and NIAK-62 are merged since all requests are rejected. 
So I have a suggestion, maybe, once NIAK-55 is fixed, it could be merged into NIAK-61 and then we could leave only NIAK-61 PR open? NIAK-62 is contained in NIAK-61 anyway if I'm correct. Then I shall review NIAK-61 with a shorter delay than I did this branch :)

Sewerinius committed:
[NIAK-55] Merge branch 'develop' into NIAK-55

# Conflicts:
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/ConnectionController.java
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/CourseOfStudyController.java

Sewerinius committed:
[NIAK-55] Secured more endpoints

Sewerinius committed:
[NIAK-55] Merge remote-tracking branch 'remotes/origin/develop' into NIAK-55

# Conflicts:
#	rezerwacje-app/config/checkstyle/checkstyle.xml
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/ConnectionController.java
#	rezerwacje-app/src/main/java/org/niewidoczniakademicy/rezerwacje/controller/RecruitmentController.java

Sewerinius committed:
[NIAK-55] Secure controllers
